### PR TITLE
Update Alpaka code base and build rules

### PR DESCRIPTION
**Pull request by fwyzard**
Try to simplify and reduce both the code size and the binary size of the Alpaka implementation:
  - move accelerator-independent types to the `alpaka_common` namespace (suggestions for a better name are welcome)
  - move accelerator-independent functions to the global namespace (outside of the `ALPAKA_ARCHITECTURE` namespaces)
  - move the kernel launch from the `rawtodigi()` wrapper to the `analyze()` function, and use explicit template specialisation to avoid having the whole kernel in the header file
  - compile individual .cc files and reuse the .o files for the different Alpaka targets
  - use separate compilation and linking of CUDA device code

### CODE REVIEW

**Comment by fwyzard** [comment]
This PR includes (and builds upon) the commits from #23 .

**Comment by fwyzard** [comment]
@makortel @waredjeb @rovere @felicepantaleo

I still have few issues with the way we are using Alpaka here (or maybe with the way Alpaka assumes the code to be written ?).


In principle I think we could build some of the "device" code (that does not depend on the actual `T_Acc` type) once per architecture (e.g. CPU vs CUDA) rather than once per backend (i.e. serial vs TBB vs CUDA).
Currently, all "device" functions are marked as `inline` to avoid duplicate symbols at compile time, but we are still compiling them multiple times.
On the other hand, the expectation is that those function should actually be templated on the actual `T_Acc` accelerator type, if they use any of the more advanced functionality (atomics, shared memory, synchronisation, etc.).
So, are we seeing this only because we are using an overly simple code base ?  If so, is it something we want so optimise for anyway ?


When using native CUDA, only the device and kernel code needs to be compiled by nvcc, the API functions and kernel launches can be compiled with the host compiler (e.g. gcc).
As far as I understand from the Alpaka implementation, that is not true with Alpaka: anything that requires the `ALPAKA_ACC_GPU_CUDA_ENABLED` symbol (so, memory copies, etc.) also requires building with nvcc.
I suspect that for what concerns the API functions, this could be relaxed, for example checking for `CUDA_VERSION` rather then for `__CUDACC__`; while for the kernel launches the `kernel<<<...>>>(...)` syntax currently used inside Alpaka does requires nvcc.
In any case, is this limitation something we should discuss with the Alpaka developers ?

**Comment by makortel** [comment]
Thanks @fwyzard. I have to digest the PR a bit, but for your question

> As far as I understand from the Alpaka implementation, that is not true with Alpaka: anything that requires the ALPAKA_ACC_GPU_CUDA_ENABLED symbol (so, memory copies, etc.) also requires building with nvcc.
> ...
> In any case, is this limitation something we should discuss with the Alpaka developers ?

I think it would be good to ask already now. I'd really like to be able to compile as much of the code with gcc (or "host compiler") as we can.

**Comment by fwyzard** [comment]
OK, I've created a minimal example that reproduces the issue of using `gcc` vs `nvcc` for the host-side API (memory operations, kernel launches) and opened an issue on the Alpaka project: https://github.com/ComputationalRadiationPhysics/alpaka/issues/908 .

File: rawtodigi_alpaka.cc [hunk 1]
```diff
@@ -1,516 +1,493 @@
...
-
-  ALPAKA_FN_HOST_ACC
-  uint32_t getLink(uint32_t ww) { return ((ww >> pixelgpudetails::LINK_shift) & pixelgpudetails::LINK_mask); }
-
-  ALPAKA_FN_HOST_ACC
-  uint32_t getRoc(uint32_t ww) { return ((ww >> pixelgpudetails::ROC_shift) & pixelgpudetails::ROC_mask); }
-
-  ALPAKA_FN_HOST_ACC
-  uint32_t getADC(uint32_t ww) { return ((ww >> pixelgpudetails::ADC_shift) & pixelgpudetails::ADC_mask); }
+class Packing {
```
Comment (makortel) [inline rawtodigi_alpaka.cc]: Would it work if all the file-scope class(es) and functions would be put into an anonymous namespace?

File: rawtodigi_alpaka.h [hunk 2]
```diff
@@ -2,19 +2,25 @@
...
+                                  const Input* input,
+                                  Output* output,
+                                  bool useQualityInfo,
+                                  bool includeErrors,
+                                  bool debug) const;
+  };
+
+  // explicit template instantiation declaration for ALPAKA_ARCHITECTURE::Acc
+  extern template ALPAKA_FN_ACC void rawtodigi_kernel::operator()(
+      Acc const& acc, const Input* input, Output* output, bool useQualityInfo, bool includeErrors, bool debug) const;
```
Comment (makortel) [inline rawtodigi_alpaka.h]: So the trick is that the caller (`analyzer_alpaka.cc`) is compiled separately for each backend, as is the implementation file (`rawtodigi_alpaka.cc`), and in `rawtodigi_alpaka.cc` the `ALPAKA_ARCHITECTURE::Acc` type drives the instantiation. The setup works as long as we compile the caller and the implementation for the same backend, and link them together. Did I get it right?

Do you have any other reason in mind for this separation of caller and implementation (i.e. not #including the implementation into the caller) than the hope for compiling the caller with the host compiler?

**Comment by makortel** [comment]
@fwyzard 
> So, are we seeing this only because we are using an overly simple code base ? If so, is it something we want so optimise for anyway ?

With "this" do you refer to "compiling `T_Acc`-independent code once per architecture instead of once per backend"? Would that be relevant for anything else than CPU architecture?

(ok, that makes me wonder what would happen with SYCL: one SYCL backend that would be compiled for multiple architectures, or multiple architecture-specific SYCL backends)

I would imagine that for each architecture we would define one backend that we would use (likely centrally, or possibly per use-case if that customization would turn out to be important).

**Comment by makortel** [comment]
It came to my mind that Kokkos has a similar constraint
> Any Kokkos application compiled for CUDA embeds CUDA code via template metaprogramming. Thus, the whole application must be built with a CUDA-capable compiler. (At the moment, the only such compilers are NVIDIA's NVCC and Clang 4.0.) More precisely, every compilation unit containing a Kokkos kernel or a function called from a Kokkos kernel has to be compiled with a CUDA-capable compiler. This includes files containing Kokkos::View allocations which call an initialization kernel.

https://github.com/kokkos/kokkos/wiki/Compiling#44-building-for-cuda

**Comment by fwyzard** [comment]
Rebased on top of `master`.

File: rawtodigi_alpaka.cc (hunk 1 above)
Comment (fwyzard) [reply rawtodigi_alpaka.cc]: Yes.

File: rawtodigi_alpaka.h (hunk 2 above)
Comment (fwyzard) [reply rawtodigi_alpaka.h]: > Did I get it right?

Correct.

> Do you have any other reason in mind for this separation of caller and implementation (i.e. not #including the implementation into the caller) than the hope for compiling the caller with the host compiler?

No, that is the main reason, apart from the usual reason of trying to keep the individual files to manageable size.

**Comment by fwyzard** [comment]
> > So, are we seeing this only because we are using an overly simple code base ? If so, is it something we want so optimise for anyway ?
> 
> With "this" do you refer to "compiling `T_Acc`-independent code once per architecture instead of once per backend"?

Correct.

> Would that be relevant for anything else than CPU architecture?

If you mean, do I expect there to be other hardware architectures with multiple Alpaka backends ? No, I can't think of any right now.

If you mean, does leaving out both the namespace and template affect other architectures ?
I don't know. 

For CUDA vs CPU it works because the compiler knows how to handle `__host__` vs `__device__` functions anyway. I'm not sure what happens if we add HIP to the mix.

> (ok, that makes me wonder what would happen with SYCL: one SYCL backend that would be compiled for multiple architectures, or multiple architecture-specific SYCL backends)

Do you mean SYCL by itself, or as a backend for Alpaka ?
Either way thinking about it makes my brain hurt :-(

> I would imagine that for each architecture we would define one backend that we would use (likely centrally, or possibly per use-case if that customization would turn out to be important).

Leaving SYCL out, I think the only architecture for which having two backends makes sense is the CPU: a serial backend, and a TBB/OpenMP backend to implement intra-module parallelism.

I guess if we end up doing it, we can always define one "architecture namespace" and one "backend namespace", and have different backends (e.g. serial and TBB) share the same architecture namespace.

**Comment by fwyzard** [comment]
>  ... we can always define one "architecture namespace" and one "backend namespace", and have different backends (e.g. serial and TBB) share the same architecture namespace.

Shall I give it a try ?

**Comment by makortel** [comment]
> > Would that be relevant for anything else than CPU architecture?
> 
> If you mean, do I expect there to be other hardware architectures with multiple Alpaka backends ? No, I can't think of any right now.

This is what I had in mind.

> If you mean, does leaving out both the namespace and template affect other architectures ?
> I don't know.

I didn't think of dropping both. I'd believe we would in general want at least one of (namespace, template) to distinguish the symbols, even if we'd try hard to never load non-CPU shared objects for more than one architecture.

> For CUDA vs CPU it works because the compiler knows how to handle `__host__` vs `__device__` functions anyway. I'm not sure what happens if we add HIP to the mix.

Right.

> > (ok, that makes me wonder what would happen with SYCL: one SYCL backend that would be compiled for multiple architectures, or multiple architecture-specific SYCL backends)
> 
> Do you mean SYCL by itself, or as a backend for Alpaka ?

I meant SYCL as a backend for Alpaka.

> Either way thinking about it makes my brain hurt :-(

Indeed.

> > I would imagine that for each architecture we would define one backend that we would use (likely centrally, or possibly per use-case if that customization would turn out to be important).
> 
> Leaving SYCL out, I think the only architecture for which having two backends makes sense is the CPU: a serial backend, and a TBB/OpenMP backend to implement intra-module parallelism.

I had thought we'd use the TBB backend also for the serial case (even if it incurs a runtime penalty, which I hope to be tiny).

> I guess if we end up doing it, we can always define one "architecture namespace" and one "backend namespace", and have different backends (e.g. serial and TBB) share the same architecture namespace.

At this point I'm not sure if the added complexity would really be worth it.

**Comment by fwyzard** [comment]
> At this point I'm not sure if the added complexity would really be worth it.

I've made the changes with the latest commit, it doesn't seem to add *too much* complexity... and it should be easy to remove if we don't like it.

File: alpakaConfig.h [hunk 3]
```diff
@@ -3,58 +3,65 @@
...
   using Queue = alpaka::queue::QueueCudaRtNonBlocking;
-  using WorkDiv = alpaka::workdiv::WorkDivMembers<Dim, Idx>;
-  using Vec = alpaka::vec::Vec<Dim, Idx>;
 }  // namespace alpaka_cuda_async
 
 #endif  // ALPAKA_ACC_GPU_CUDA_ENABLED
 
+#ifdef ALPAKA_ACC_GPU_CUDA_BACKEND
+#define ALPAKA_ARCHITECTURE_NAMESPACE alpaka_cuda
+#define ALPAKA_ACCELERATOR_NAMESPACE alpaka_cuda_async
```
Comment (makortel) [inline alpakaConfig.h:30]: How would you feel about calling the namespace `ALPAKA_BACKEND_NAMESPACE`? I'm just thinking that `ACCELERATOR` sounds a bit confusing for CPU architecture. On the other hand Alpaka calls all backends as "accelerators", so maybe `ACCELERATOR` is fine.

Comment (fwyzard) [reply alpakaConfig.h:30]: I picked `ALPAKA_ACCELERATOR_NAMESPACE` to mirror Cupla that uses `CUPLA_ACCELERATOR_NAMESPACE`, but I don't really mind the name.

Comment (makortel) [reply alpakaConfig.h:30]: Would it be useful to use a consistent naming? (obviously can be changed later with minor cost, so not a big issue)

Comment (makortel) [reply alpakaConfig.h:30]: Let's stay with the `ACCELERATOR` for now.

**Pull request merged**